ouroboros-network/src/Ouroboros/Network/Subscription/Subscriber.hs

```diff
@@ -0,0 +1,26 @@
+-- | Definition of a modular subscriber action, which maintains subscriptions
+-- through failures etc.
+
+module Ouroboros.Network.Subscription.Subscriber
+    ( SubscriptionTarget (..)
+    , constantSubscriptionTarget
+    , listSubscriptionTarget
+    ) where
+
+-- | Generate subscription targets in some monad.
+-- TBD any value in using a streaming solution like conduit?
+newtype SubscriptionTarget m target = SubscriptionTarget
+    { getSubscriptionTarget :: m (Maybe (target, SubscriptionTarget m target))
+    }
+
+constantSubscriptionTarget :: Applicative m => target -> SubscriptionTarget m target
+constantSubscriptionTarget target =
+    SubscriptionTarget (pure (Just (target, constantSubscriptionTarget target)))
+
+listSubscriptionTarget
+    :: Applicative m
+    => [target]
+    -> SubscriptionTarget m target
+listSubscriptionTarget []     = SubscriptionTarget $ pure Nothing
+listSubscriptionTarget (t:ts) = SubscriptionTarget $ pure (Just (t, listSubscriptionTarget ts))
```

Comment:
I was tricked at the beginning that there is no `repeat` subscription target which repeats the given list of `ip` addressees.  It is done by `ipSubscriptionWorker`.   It is re-using `listSubscriptTarget` in a loop doing some monadic actions on start and end, I think one could fit it into a `SubscriptionTarget`.

Reply:
Such a change would have to work with `Dns.hs` too.
In `subscriptionWorker` there also a delay to prevent us from flooding the network with continues connection attempts towards a peer that is down.

For now I think the `forever` in `subscriptionWorker` is fine.